### DISCUSSION TITLE
No point detected during inference

### DISCUSSION THREAD

**Discussion by edofazza** [talmolab/sleap#814] [Help!]
# Bug description
When I tried to predict using the model I trained all the predicted frames got 0 keypoints detected.

## Expected behaviour
Prediction should be fine since the training of the single instance model was successful.

## Actual behaviour
Predictions are incorrect, as the image below shows. The number of points is 0/7 instead of 7/7 and this happens for all the predicted frames. I also tried to predict using an entire video and I got the same bad results even for the frames the model was trained on.

## Your personal set up
 - OS:
 ubuntu 20.04 (for the training on a remote virtual machine equipped with NVIDIA A100-SXM4 40Gb) and macOS Monterey v12.4 (for the inferring using the GUI or through code)
 - Version(s): 
SLEAP v1.2.3, python 3.8.10 (on Ubuntu), python3.9.12 (on macOS)
 - SLEAP installation method (listed [here](https://sleap.ai/installation.html#)):
    * [ ] [Conda from package](https://sleap.ai/installation.html#conda-package)
    * [x] [Conda from source](https://sleap.ai/installation.html#conda-from-source) (on macOS)
    * [x] [pip package](https://sleap.ai/installation.html#pip-package) (on Ubuntu)
    * [ ] [M1 Macs](https://sleap.ai/installation.html#m1-macs)

<details><summary>Environment packages</summary>

On Ubuntu:
absl-py==1.1.0
appdirs==1.4.3
astunparse==1.6.3
attrs==21.2.0
backports.zoneinfo==0.2.1
cachetools==5.2.0
cattrs==1.1.1
certifi==2019.11.28
chardet==3.0.4
colorama==0.4.5
commonmark==0.9.1
cycler==0.11.0
dbus-python==1.2.16
distlib==0.3.0
efficientnet==1.0.0
enum34==1.1.10
filelock==3.0.12
flatbuffers==2.0
fonttools==4.33.3
future==0.18.2
gast==0.5.3
google-auth==2.8.0
google-auth-oauthlib==0.4.6
google-pasta==0.2.0
grpcio==1.46.3
h5py==3.6.0
idna==2.8
image-classifiers==1.0.0
imageio==2.15.0
imgaug==0.4.0
imgstore==0.2.9
importlib-metadata==4.11.4
joblib==1.1.0
jsmin==3.0.1
jsonpickle==1.2
keras==2.8.0
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.2
kiwisolver==1.4.3
libclang==14.0.1
Markdown==3.3.7
matplotlib==3.5.2
mock==4.0.3
more-itertools==4.2.0
networkx==2.8.4
numpy==1.21.5
oauthlib==3.2.0
opencv-python==4.6.0.66
opencv-python-headless==4.5.5.62
opt-einsum==3.3.0
packaging==21.3
pandas==1.4.2
Pillow==9.1.1
portpicker==1.5.2
protobuf==3.19.4
psutil==5.9.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
Pygments==2.12.0
PyGObject==3.36.0
pykalman==0.9.5
pyparsing==3.0.9
PySide2==5.14.1
python-apt==2.0.0+ubuntu0.20.4.7
python-dateutil==2.8.2
python-rapidjson==1.6
pytz==2022.1
pytz-deprecation-shim==0.1.0.post0
PyWavelets==1.3.0
PyYAML==6.0
pyzmq==23.2.0
qimage2ndarray==1.8.3
requests==2.22.0
requests-oauthlib==1.3.1
requests-unixsocket==0.2.0
rich==10.16.1
rsa==4.8
scikit-image==0.19.3
scikit-learn==1.0.2
scikit-video==1.1.11
scipy==1.7.3
seaborn==0.11.2
segmentation-models==1.0.1
Shapely==1.8.2
shiboken2==5.14.1
six==1.14.0
sklearn==0.0
sleap==1.2.3
tb-nightly==2.10.0a20220622
tensorboard==2.8.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorflow==2.8.2
tensorflow-estimator==2.8.0
tensorflow-io-gcs-filesystem==0.26.0
termcolor==1.1.0
threadpoolctl==3.1.0
tifffile==2022.5.4
typing-extensions==4.2.0
tzdata==2022.1
tzlocal==4.2
urllib3==1.25.8
virtualenv==20.0.17
Werkzeug==2.1.2
wrapt==1.14.1
zipp==1.0.0

On macOS (conda env)
absl-py==1.0.0
alabaster==0.7.12
anyio==3.6.1
appdirs==1.4.4
appnope==0.1.3
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
astunparse==1.6.3
attrs==21.2.0
Babel==2.10.1
backcall==0.2.0
backports.zoneinfo==0.2.1
beautifulsoup4==4.11.1
black==21.6b0
bleach==5.0.0
cached-property @ file:///tmp/build/80754af9/cached-property_1600785575025/work
cachetools==5.2.0
cattrs==1.1.1
certifi==2021.10.8
cffi==1.15.0
cfgv==3.3.1
charset-normalizer==2.0.12
click==8.0.4
colorama==0.4.4
commonmark==0.9.1
coverage==6.4
cycler==0.11.0
debugpy==1.6.0
decorator==5.1.1
defusedxml==0.7.1
Deprecated==1.2.13
distlib==0.3.4
docutils==0.17.1
efficientnet==1.0.0
entrypoints==0.4
filelock==3.7.0
flatbuffers==2.0
fonttools==4.33.3
gast==0.5.3
gitdb==4.0.9
GitPython==3.1.27
google-auth==2.6.6
google-auth-oauthlib==0.4.6
google-pasta==0.2.0
greenlet==1.1.2
grpcio==1.46.3
h5py @ file:///Users/runner/miniforge3/conda-bld/h5py_1637964031503/work
identify==2.5.1
idna==3.3
image-classifiers==1.0.0
imageio==2.15.0
imagesize==1.3.0
imgaug==0.4.0
imgstore==0.2.9
importlib-metadata==4.11.4
importlib-resources==5.7.1
iniconfig==1.1.1
ipykernel==6.13.0
ipython==7.34.0
ipython-genutils==0.2.0
ipywidgets==7.7.0
jedi==0.17.2
Jinja2==3.1.2
joblib==1.1.0
jsmin==3.0.1
json5==0.9.8
jsonpickle==1.2
jsonschema==4.5.1
jupyter-cache==0.4.3
jupyter-client==7.3.1
jupyter-core==4.10.0
jupyter-server==1.13.5
jupyter-server-mathjax==0.2.5
jupyter-sphinx==0.3.2
jupyterlab==3.3.4
jupyterlab-pygments==0.2.2
jupyterlab-server==2.14.0
jupyterlab-widgets==1.1.0
keras==2.8.0
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.2
keyring==23.5.1
kiwisolver==1.4.2
libclang==14.0.1
linkify-it-py==2.0.0
livereload==2.6.3
Markdown==3.3.7
markdown-it-py==1.1.0
MarkupSafe==2.1.1
matplotlib==3.5.2
matplotlib-inline==0.1.3
mdit-py-plugins==0.2.8
mistune==0.8.4
mypy-extensions==0.4.3
myst-nb==0.13.2
myst-parser==0.15.2
nbclassic==0.3.7
nbclient==0.5.13
nbconvert==6.5.0
nbdime==3.1.1
nbformat==5.1.3
nest-asyncio==1.5.5
networkx==2.6.3
nodeenv==1.6.0
notebook==6.4.11
notebook-shim==0.1.0
numpy @ file:///Users/runner/miniforge3/conda-bld/numpy_1649585296287/work
oauthlib==3.2.0
olefile==0.46
opencv-python==4.5.5.64
opencv-python-headless==4.5.5.62
opt-einsum==3.3.0
packaging==21.3
pandas==1.3.5
pandocfilters==1.5.0
parso==0.7.1
pathspec==0.9.0
pexpect==4.8.0
pickleshare==0.7.5
Pillow==8.4.0
pkginfo==1.8.2
platformdirs==2.5.2
pluggy==1.0.0
pre-commit==2.19.0
prometheus-client==0.14.1
prompt-toolkit==3.0.29
protobuf==3.19.4
psutil==5.9.1
ptyprocess==0.7.0
py==1.11.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.21
pydata-sphinx-theme==0.8.1
PyGithub==1.55
Pygments==2.12.0
PyJWT==2.4.0
pykalman==0.9.5
PyNaCl==1.5.0
pyparsing==3.0.9
pyrsistent==0.18.1
PySide2==5.13.2
pytest==7.1.2
pytest-cov==3.0.0
pytest-qt==4.0.2
pytest-xvfb==2.0.0
python-dateutil @ file:///tmp/build/80754af9/python-dateutil_1626374649649/work
python-rapidjson==1.6
pytz==2021.3
pytz-deprecation-shim==0.1.0.post0
PyVirtualDisplay==3.0
PyWavelets==1.3.0
PyYAML==6.0
pyzmq==23.0.0
qimage2ndarray==1.8.3
readme-renderer==35.0
regex==2022.4.24
requests==2.27.1
requests-oauthlib==1.3.1
requests-toolbelt==0.9.1
rfc3986==2.0.0
rich==10.16.1
rsa==4.8
scikit-image==0.19.2
scikit-learn==1.0.2
scikit-video==1.1.11
scipy @ file:///Users/runner/miniforge3/conda-bld/scipy_1637806768271/work
seaborn==0.11.2
segmentation-models==1.0.1
Send2Trash==1.8.0
Shapely @ file:///opt/concourse/worker/volumes/live/188867f5-a412-4cde-6476-bb38afac2520/volume/shapely_1632813489757/work
shiboken2==5.13.2
six @ file:///tmp/build/80754af9/six_1644875935023/work
-e git+https://github.com/talmolab/sleap@293126219c688f71f5045be13aacb424318cb557#egg=sleap
smmap==5.0.0
sniffio==1.2.0
snowballstemmer==2.2.0
soupsieve==2.3.2.post1
Sphinx==4.5.0
sphinx-autobuild==2021.3.14
sphinx-book-theme==0.3.2
sphinx-copybutton==0.5.0
sphinx-togglebutton==0.3.1
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==2.0.0
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.5
SQLAlchemy==1.4.36
tensorboard==2.8.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorflow==2.8.2
tensorflow-estimator==2.8.0
tensorflow-io-gcs-filesystem==0.26.0
termcolor==1.1.0
terminado==0.15.0
threadpoolctl==3.1.0
tifffile==2021.11.2
tinycss2==1.1.1
toml==0.10.2
tomli==2.0.1
tornado==6.1
tqdm==4.64.0
traitlets==5.2.1.post0
twine==3.3.0
typed-ast==1.5.4
typing_extensions==4.2.0
tzdata==2022.1
tzlocal==4.2
uc-micro-py==1.0.1
urllib3==1.26.9
virtualenv==20.14.1
wcwidth==0.2.5
webencodings==0.5.1
websocket-client==1.3.2
Werkzeug==2.1.2
widgetsnbextension==3.6.0
wrapt==1.14.1
zipp==3.8.0
</details>

<details><summary>Logs</summary>
Using already trained model for single_instance: /Users/edoardo/Desktop/crickets/remote tests/models/220623_145216.single_instance/training_config.json
Command line call:
sleap-track /Users/edoardo/Desktop/crickets/cropped/ref_reduced/C1_A_0137.mp4 --frames 0,163,7,72,47,111,145,81,181,150,88,156,61,62 -m /Users/edoardo/Desktop/crickets/remote tests/models/220623_145216.single_instance/training_config.json --tracking.tracker none -o /Users/edoardo/Desktop/crickets/predictions/C1_A_0137.mp4.220624_101524.predictions.slp --verbosity json --no-empty-frames

Started inference at: 2022-06-24 10:15:41.078114
Args:
{
│   'data_path': '/Users/edoardo/Desktop/crickets/cropped/ref_reduced/C1_A_0137.mp4',
│   'models': [
│   │   '/Users/edoardo/Desktop/crickets/remote tests/models/220623_145216.single_instance/training_config.json'
│   ],
│   'frames': '0,163,7,72,47,111,145,81,181,150,88,156,61,62',
│   'only_labeled_frames': False,
│   'only_suggested_frames': False,
│   'output': '/Users/edoardo/Desktop/crickets/predictions/C1_A_0137.mp4.220624_101524.predictions.slp',
│   'no_empty_frames': True,
│   'verbosity': 'json',
2022-06-24 10:15:41.722032: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
│   'video.dataset': None,
│   'video.input_format': 'channels_last',
│   'cpu': False,
│   'first_gpu': False,
│   'last_gpu': False,
│   'gpu': 0,
│   'max_edge_length_ratio': 0.25,
│   'dist_penalty_weight': 1.0,
│   'batch_size': 4,
│   'open_in_gui': False,
│   'peak_threshold': 0.2,
│   'tracking.tracker': 'none',
│   'tracking.target_instance_count': None,
│   'tracking.pre_cull_to_target': None,
│   'tracking.pre_cull_iou_threshold': None,
│   'tracking.post_connect_single_breaks': None,
│   'tracking.clean_instance_count': None,
│   'tracking.clean_iou_threshold': None,
│   'tracking.similarity': None,
│   'tracking.match': None,
│   'tracking.track_window': None,
│   'tracking.min_new_track_points': None,
│   'tracking.min_match_points': None,
│   'tracking.img_scale': None,
│   'tracking.of_window_size': None,
│   'tracking.of_max_levels': None,
│   'tracking.kf_node_indices': None,
│   'tracking.kf_init_frame_count': None
}
2022-06-24 10:15:54.614782: W tensorflow/core/grappler/costs/op_level_cost_estimator.cc:690] Error in PredictCost() for the op: op: "CropAndResize" attr { key: "T" value { type: DT_FLOAT } } attr { key: "extrapolation_value" value { f: 0 } } attr { key: "method" value { s: "bilinear" } } inputs { dtype: DT_FLOAT shape { dim { size: -29 } dim { size: -30 } dim { size: -31 } dim { size: 1 } } } inputs { dtype: DT_FLOAT shape { dim { size: -12 } dim { size: 4 } } } inputs { dtype: DT_INT32 shape { dim { size: -12 } } } inputs { dtype: DT_INT32 shape { dim { size: 2 } } } device { type: "CPU" vendor: "GenuineIntel" model: "110" frequency: 3000 num_cores: 6 environment { key: "cpu_instruction_set" value: "SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2" } environment { key: "eigen" value: "3.4.90" } l1_cache_size: 32768 l2_cache_size: 262144 l3_cache_size: 9437184 memory_size: 268435456 } outputs { dtype: DT_FLOAT shape { dim { size: -12 } dim { size: -32 } dim { size: -33 } dim { size: 1 } } }

Versions:
SLEAP: 1.2.3
TensorFlow: 2.8.2
Numpy: 1.21.5
Python: 3.7.13
OS: Darwin-21.5.0-x86_64-i386-64bit

System:
GPUs: None detected.

Video: /Users/edoardo/Desktop/crickets/cropped/ref_reduced/C1_A_0137.mp4
2022-06-24 10:16:50.239263: W tensorflow/core/grappler/costs/op_level_cost_estimator.cc:690] Error in PredictCost() for the op: op: "CropAndResize" attr { key: "T" value { type: DT_FLOAT } } attr { key: "extrapolation_value" value { f: 0 } } attr { key: "method" value { s: "bilinear" } } inputs { dtype: DT_FLOAT shape { dim { size: -29 } dim { size: -30 } dim { size: -31 } dim { size: 1 } } } inputs { dtype: DT_FLOAT shape { dim { size: -12 } dim { size: 4 } } } inputs { dtype: DT_INT32 shape { dim { size: -12 } } } inputs { dtype: DT_INT32 shape { dim { size: 2 } } } device { type: "CPU" vendor: "GenuineIntel" model: "110" frequency: 3000 num_cores: 6 environment { key: "cpu_instruction_set" value: "SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2" } environment { key: "eigen" value: "3.4.90" } l1_cache_size: 32768 l2_cache_size: 262144 l3_cache_size: 9437184 memory_size: 268435456 } outputs { dtype: DT_FLOAT shape { dim { size: -12 } dim { size: -32 } dim { size: -33 } dim { size: 1 } } }
Finished inference at: 2022-06-24 10:16:55.978656
Total runtime: 74.90057110786438 secs
Predicted frames: 14/14
Provenance:
{
│   'sleap_version': '1.2.3',
│   'platform': 'Darwin-21.5.0-x86_64-i386-64bit',
│   'command': '/Users/edoardo/miniconda3/envs/sleap/bin/sleap-track /Users/edoardo/Desktop/crickets/cropped/ref_reduced/C1_A_0137.mp4 --frames 0,163,7,72,47,111,145,81,181,150,88,156,61,62 -m /Users/edoardo/Desktop/crickets/remote tests/models/220623_145216.single_instance/training_config.json --tracking.tracker none -o /Users/edoardo/Desktop/crickets/predictions/C1_A_0137.mp4.220624_101524.predictions.slp --verbosity json --no-empty-frames',
│   'data_path': '/Users/edoardo/Desktop/crickets/cropped/ref_reduced/C1_A_0137.mp4',
│   'model_paths': [
│   │   '/Users/edoardo/Desktop/crickets/remote tests/models/220623_145216.single_instance/training_config.json'
│   ],
│   'output_path': '/Users/edoardo/Desktop/crickets/predictions/C1_A_0137.mp4.220624_101524.predictions.slp',
│   'predictor': 'SingleInstancePredictor',
│   'total_elapsed': 74.90057110786438,
│   'start_timestamp': '2022-06-24 10:15:41.078114',
│   'finish_timestamp': '2022-06-24 10:16:55.978656'
}

Saved output: /Users/edoardo/Desktop/crickets/predictions/C1_A_0137.mp4.220624_101524.predictions.slp

Process return code: 0
</details>

## Screenshots
<img width="1476" alt="Schermata 2022-06-24 alle 10 40 21" src="https://user-images.githubusercontent.com/46302551/175498146-2df936c1-06c7-481a-8027-a2ff7f4ab010.png">

## How to reproduce
There are two ways to reproduce the error. One is
1. Run `sleap-label`
2. Open project
3. Select `predict` -> `run inference`
4. Load single instance model
5. Run inference (e.g., on a random sample)  

The other one is using the following code:
```
import sleap

if __name__ == '__main__':
    sleap.disable_preallocation()

    video = sleap.load_video('test.mp4')
    print(video.shape)

    predictor = sleap.load_model('/models/220623_145216.single_instance')
    labels = predictor.predict(video)

    labels = sleap.Labels(labels.labeled_frames[:100])  # using only the first 100 frames
    labels.save("labels_with_images_p.pkg.slp", with_images=True, embed_all_labeled=True)
```

P.S: I will send you my data to test the issue at talmo@salk.edu

**Comment by roomrys** [comment]
Tagging similar behavior here (Inference unable to place skeleton #812)

**Comment by roomrys** [comment]
Hi @edofazza,

I took a look at the data you sent and believe it is the [model settings](https://github.com/talmolab/sleap/discussions/812) that are causing you trouble. I have changed the settings slightly to consistently predict the tip of the head (by setting the RF to cover the antennae and also changing the min/max rotation to +-180). 

I believe the rest of the antennae are now being missed due to losing resolution when decreasing the input scaling (which we do to increase the RF size). I am running a few trials at the moment to check how high the input scaling can be to get the results we want. Will keep you updated on the results.

![image](https://user-images.githubusercontent.com/38435167/175663694-877fe950-d1c3-44ad-bdaa-8ad2dbb3a584.png)
**Fig 1: Just the head is detected, likely to lower resolution with lower input scaling.**

![image](https://user-images.githubusercontent.com/38435167/175663821-4faa5040-27ed-4f53-887d-bb551c0c207f.png)
**Fig 2: New model parameters that detect just the head. Higher resolution may be needed to detect the antennae.**

Thanks,
Liezl
